### PR TITLE
bpo-17611. Move unwinding of stack for "pseudo exceptions" from interpreter to compiler.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue17611

### CODE REVIEW

File: Python/compile.c
Comment (pitrou) [inline Python/compile.c]: This forward decl doesn't look necessary.

File: Python/compile.c
Comment (pitrou) [inline Python/compile.c]: This doesn't match the END_FINALLY description in `dis.rst` above.

**Comment by pitrou** [comment]
Thanks for doing this! I'm impressed that you managed to reduce the line count of `frameobject.c` and moderate the code growth in `compile.c`.
* Did you run any benchmarks?
* Please merge git master into this branch, to get the new line tracing tests

File: Lib/test/test_compile.py
Comment (pitrou) [inline Lib/test/test_compile.py]: Do you have any idea how to solve this one?

**Comment by nascheme** [comment]
The SETUP_FINALLY/END_FINALLY subroutine idea looks pretty clean to me.  Simplifies the compiler as the fblock stack is no longer a tree (as in my PR).

Failing test:

    # crashes 9273f101759b9df4198b4aad6d24bc11c2c3bca7
    # python: ../Python/ceval.c:1881: _PyEval_EvalFrameDefault: Assertion `STACK_LEVEL() >= (b)->b_level + 3 && STACK_LEVEL() <= (b)->b_level + 4' failed.
    def f():
        try:
            raise Exception
        except:
            try:
                x = 1
            finally:
                return 1
    f()


Another:

    # crash c9273f101759b9df4198b4aad6d24bc11c2c3bca7
    # python: ../Python/ceval.c:1686: _PyEval_EvalFrameDefault: Assertion `f->f_iblock == 0' failed.
    def func():
        try:
            raise RuntimeError
        finally:
            print('finally 1')
            return 1
    func()

**Comment by serhiy-storchaka** [comment]
Good catch @nascheme! Unwinding don't work in a finally block still. :(

File: Lib/test/test_compile.py
Comment (serhiy-storchaka) [reply Lib/test/test_compile.py]: No, I don't have any idea. But since this test is failed with the current code I don't worry. Maybe bpo-31113 will fix it.

**Comment by serhiy-storchaka** [comment]
Fixed @nascheme's examples by adding a new opcode `POP_FINALLY`. It is not documented yet since I'm not sure what approach is better.

I didn't run any benchmarks yet.

**Comment by pitrou** [comment]
Do you plan to add tests for @nascheme's examples?

**Comment by serhiy-storchaka** [comment]
Yes, once I found a good place for them. And if we don't decide to prohibit 'break' and 'return' in 'finally' clause ('continue' already is prohibited).

**Comment by nascheme** [comment]
There is certainly code "in the wild" that uses 'return' inside a 'finally' clause.  I think greenlet or gevent was one example.

File: Lib/opcode.py
Comment (pitrou) [inline Lib/opcode.py]: Why is POP_FINALLY a relative jump?

File: Python/compile.c
Comment (pitrou) [inline Python/compile.c]: Why is this statically -6? This would deserve a comment.

File: Doc/library/dis.rst
Comment (pitrou) [inline Doc/library/dis.rst:689]: For the sake of static bytecode analyzers, perhaps it would be better to have two separate opcodes: one preserving TOS, the other not.

Comment (serhiy-storchaka) [reply Doc/library/dis.rst:689]: Wouldn't this just complicate the code and the documentation?

File: Lib/opcode.py
Comment (serhiy-storchaka) [reply Lib/opcode.py]: Good catch! This is an error.

File: Python/compile.c
Comment (serhiy-storchaka) [reply Python/compile.c]: Done.

File: Doc/library/dis.rst
Comment (pitrou) [reply Doc/library/dis.rst:689]: The code wouldn't be much more complicated (you can share the same implementation in the eval loop). OTOH I'm not sure it's necessary.

Other question: why are both POP_FINALLY and END_FINALLY needed? The documentation should be a bit clearer.

Comment (serhiy-storchaka) [reply Doc/library/dis.rst:689]: `END_FINALLY` re-raises an exception, `POP_FINALLY 0` just drops it. `END_FINALLY` jumps to the return address, `POP_FINALLY 0` ignores it. And `POP_FINALLY 1` preserves TOS.

I don't know what to add to the documentation. Can you provide your changes?

If push 6 `None` instead of 1 `NULL` in `BEGIN_FINALLY`,  and push an exception block in `BEGIN_FINALLY`, and update other instructions accordingly, and add additional hacks to stack effect calculation, `POP_FINALLY 0` could be replaced with:
```
POP_TOP
POP_TOP
POP_TOP
POP_EXCEPT
```
and `POP_FINALLY 1` could be replaced with:
```
ROT_FOUR
POP_TOP
POP_TOP
POP_TOP
ROT_FOUR
POP_EXCEPT
```
This was my initial solution for @nascheme's examples. But this would add too much unneeded overhead and complication. The single new opcode has lesser cost.

Comment (pitrou) [reply Doc/library/dis.rst:689]: > END_FINALLY re-raises an exception, POP_FINALLY 0 just drops it. END_FINALLY jumps to the return address, POP_FINALLY 0 ignores it. 

Yes... So what I don't understand is why both are needed. In which case do you need to drop an exception?

Comment (serhiy-storchaka) [reply Doc/library/dis.rst:689]: 'break' and 'return' in 'finally' clause.

'continue' is specially prohibited, but it can be allowed too after merging this PR.

Comment (pitrou) [reply Doc/library/dis.rst:689]: Thank you. Can you add some explanation of this in the docs?

File: Doc/library/dis.rst
Comment (pitrou) [inline Doc/library/dis.rst]: We don't seem to be adding `versionadded` for other new opcodes, do we?

File: Doc/library/dis.rst
Comment (pitrou) [inline Doc/library/dis.rst]: I would say "but doesn't change the bytecode counter nor raise an exception".

File: Python/ceval.c
Comment (pitrou) [inline Python/ceval.c]: Explain this is in a comment here?

File: Python/ceval.c
Comment (pitrou) [inline Python/ceval.c]: I would expect a similar comment as at the to top of END_FINALLY here.

File: Python/compile.c
Comment (pitrou) [inline Python/compile.c]: Still, it pushes a NULL... ?

**Review by pitrou — DISMISSED** [review]

File: Doc/library/dis.rst
Comment (serhiy-storchaka) [reply Doc/library/dis.rst:689]: Done.

File: Doc/library/dis.rst
Comment (serhiy-storchaka) [reply Doc/library/dis.rst]: Thanks!

File: Doc/library/dis.rst
Comment (serhiy-storchaka) [reply Doc/library/dis.rst]: `versionadded` is always added for new opcodes. Do you know cases where it was omitted?

File: Python/ceval.c
Comment (serhiy-storchaka) [reply Python/ceval.c]: Done.

File: Python/ceval.c
Comment (serhiy-storchaka) [reply Python/ceval.c]: Done.

File: Python/compile.c
Comment (serhiy-storchaka) [reply Python/compile.c]: Different approach is used now. The meaning of BEGIN_FINALLY still is not trivial, but it is now explained in a comment.

File: Doc/library/dis.rst
Comment (pitrou) [reply Doc/library/dis.rst]: See `SETUP_ASYNC_WITH`. Or the 2.7 docs, which almost always omit it.

Comment (serhiy-storchaka) [reply Doc/library/dis.rst]: I think it should be added for `SETUP_ASYNC_WITH`. It is used with other new opcodes added before and after `SETUP_ASYNC_WITH`.

**Comment by serhiy-storchaka** [comment]
I think I have addressed all comments. The code is updated to 3.8, all tests are passes. What should I do for moving this PR forward?

**Comment by pitrou** [comment]
I suppose you need a review? Given the size of the PR it's not easy to find time for it...

**Comment by serhiy-storchaka** [comment]
Yes, even after simplifications and merging some changes separately the PR still is too complex.

Could you please make at least a quick review? I think we are agree with the design, the known bugs was fixed, and if there are some minor bugs are left, we will have time to catch them in live code.

Merging this PR will unblock other changes: cleaning up `with`-related opcodes (#5112 or like) and allowing 'continue' in 'finally' clause ([bpo-32489](https://bugs.python.org/issue32489), https://github.com/serhiy-storchaka/cpython/pull/2).

**Comment by pitrou** [comment]
Yes, I'm gonna make a quick review ;-)

**Review by pitrou — APPROVED** [review]
Ok, I haven't looked very closely this time, but declare this to look ok :-)

**Comment by serhiy-storchaka** [comment]
Thanks Antoine!

**Pull request merged**
